src/test/firebase-resource-manager.spec.ts

```diff
@@ -0,0 +1,420 @@
+          PARENT_RESOURCE
+        )
+      ).to.deep.equal({ projectId: PROJECT_ID });
+    });
+
+    it("should reject if Cloud project creation fails", async () => {
+      const expectedError = new Error("HTTP Error 404: Not Found");
+      mockApi
+        .expects("request")
+        .withArgs("POST", "/v1/projects", {
+          auth: true,
+          origin: api.resourceManagerOrigin,
+          timeout: 15000,
+          data: { projectId: PROJECT_ID, name: PROJECT_NAME, parent: PARENT_RESOURCE },
+        })
+        .once()
+        .rejects(expectedError);
+      mockOraWrapper.expects("start").exactly(1);
+      mockOraWrapper.expects("succeed").never();
+      mockOraWrapper.expects("fail").exactly(1);
+
+      let err;
+      try {
+        await firebaseResourceManager.createFirebaseProject(
+          PROJECT_ID,
+          PROJECT_NAME,
+          PARENT_RESOURCE
+        );
+      } catch (e) {
+        err = e;
+      }
+      expect(err.message).to.equals(
+        "Failed to create Google Cloud project. See firebase-debug.log for more info."
+      );
+      expect(err.original).to.equal(expectedError);
+    });
+
+    it("should reject if Cloud project creation polling returns response with error", async () => {
+      const expectedError = new Error("Entity already exists");
+      expectedResolvedValues = [
+        { body: { name: "operations/cp.111111111111111" } },
+        { error: expectedError },
+      ];
+      mockApi
+        .expects("request")
+        .withArgs("POST", "/v1/projects", {
+          auth: true,
+          origin: api.resourceManagerOrigin,
+          timeout: 15000,
+          data: { projectId: PROJECT_ID, name: PROJECT_NAME, parent: PARENT_RESOURCE },
+        })
+        .once()
+        .resolves(expectedResolvedValues[0]);
+      mockPoller
+        .expects("poll")
+        .withArgs({
+          pollerName: "Project Creation Poller",
+          apiOrigin: api.resourceManagerOrigin,
+          apiVersion: "v1",
+          operationResourceName: expectedResolvedValues[0].body!.name,
+        })
+        .once()
+        .resolves(expectedResolvedValues[1]);
+      mockOraWrapper.expects("start").exactly(1);
+      mockOraWrapper.expects("succeed").never();
+      mockOraWrapper.expects("fail").exactly(1);
+
+      let err;
+      try {
+        await firebaseResourceManager.createFirebaseProject(
+          PROJECT_ID,
+          PROJECT_NAME,
+          PARENT_RESOURCE
+        );
+      } catch (e) {
+        err = e;
+      }
+      expect(err.message).to.equals(
+        "Failed to create Google Cloud project. See firebase-debug.log for more info."
+      );
+      expect(err.original).to.equal(expectedError);
+    });
+
+    it("should reject if Cloud project creation polling throws error", async () => {
+      const expectedError = new Error("Timed out");
+      const expectedResolvedValue = { body: { name: "operations/cp.111111111111111" } };
+      mockApi
```

Comment:
It looks like a lot of your `mockApi.expects("request").withArgs(...)` call chains are very similar in these tests. I think the code would be more readable if you made a few helper methods like `mockCreateCloudProjectApi()` and `mockAddFirebaseApi()` which return something like this:

```ts
return mockApi
  .expects("request")
  .withArgs("POST", "/v1/projects", {
    auth: true,
    origin: api.resourceManagerOrigin,
    timeout: 15000,
    data: { projectId: PROJECT_ID, name: PROJECT_NAME, parent: PARENT_RESOURCE },
  })
  .once()
```

Then you could use:

```ts
mockCreateCloudProjectApi().resolves(expectedResolvedValue);
```

Reply:
Updated unit tests